File: Random Problems/CharReplacement/charReplacement.py

# O(26*n) time complexity
# Dictionary and sliding window technique

def characterReplacement(s: str, k: int):        
        res=0
        count={}  # Dictionary for character count
        l=0
        for r in range(len(s)):
            #updating count for character
            count[s[r]]=1+count.get(s[r],0) 
            
            if (r-l+1)-max(count.values())>k: 
                count[s[l]]-=1
                l+=1

            # Maximum window length
            res=max(res,r-l+1) 

        return res

# Driver code
if __name__ == "__main__":
    input_str=input()
    k=int(input()) #maximum replacements
    print(characterReplacement(input_str,k))

File: Random Problems/CharReplacement/CharReplacement.cpp
#include<bits/stdc++.h>
using namespace std;
int characterReplacement(string s, int k) {
    int res = 0;
    unordered_map<char, int> count;
    int maxF = 0;

    int l = 0;

    for (int i = 0; i < s.length(); i++) {
        if(count.find(s[i]) != count.end()) 
            count[s[i]]+=1;
            
        else
            count[s[i]] = 1;
            

        maxF = max(maxF, count[s[i]]);
       
            if (((i - l + 1) - maxF) > k)
            {
                count[s[l]] -= 1;
                l++;
            }
        res = max(res, (i - l + 1));
    }
    
    return res;
}
int main(){

    string str;cin >> str;
    int k;cin>>k;
    cout<<characterReplacement(str, k)<<endl;
}
